### ISSUE TITLE
Question about the entropy term in the policy loss

### ISSUE THREAD

**Issue by niamorg** [nicklashansen/tdmpc2#76]
Hi, I am wondering about the entropy term in the policy loss in TDMPC2.

In `update_pi()`, the policy loss is defined as:

`pi_loss = (-(self.cfg.entropy_coef * info["scaled_entropy"] + qs).mean(dim=(1,2)) * rho).mean()`

The quantity `info["scaled_entropy"]` is defined in the `pi()` method of the `WorldModel` class as:

`-log_prob * entropy_scale` where `log_prob` is the log probability of the squashed gaussian policy and 
`entropy_scale = scaled_log_prob / (log_prob + 1e-8)` where `scaled_log_prob = log_prob * size` (in this last equation `log_prob` is the log probability of the NON-squashed gaussian policy (the `log_prob` variable is mutated into the squashed log prob after defining `scaled_log_prob`).

Therefore we have, by ignoring the 1e-8 term: `info["scaled_entropy"] = -squashed_log_prob * (non_squashed_log_prob * size) /  (squashed_log_prob + 1e-8) =  non_squashed_log_prob * size`.

So in the entropy term is `-entropy_coef  * non_squashed_log_prob * size`.

I am wondering why this expression instead of the regular SAC term `-entropy_coef  * squashed_log_prob * size`. Also, why is it computed in such a convoluted manner, did I miss anything?

The `WorldModel.pi()` function below for reference:

```
def pi(self, z, task):
	"""
	Samples an action from the policy prior.
	The policy prior is a Gaussian distribution with
	mean and (log) std predicted by a neural network.
	"""
	if self.cfg.multitask:
		z = self.task_emb(z, task)

	# Gaussian policy prior
	mean, log_std = self._pi(z).chunk(2, dim=-1)
	log_std = math.log_std(log_std, self.log_std_min, self.log_std_dif)
	eps = torch.randn_like(mean)

	if self.cfg.multitask: # Mask out unused action dimensions
		mean = mean * self._action_masks[task]
		log_std = log_std * self._action_masks[task]
		eps = eps * self._action_masks[task]
		action_dims = self._action_masks.sum(-1)[task].unsqueeze(-1)
	else: # No masking
		action_dims = None

	log_prob = math.gaussian_logprob(eps, log_std)

	# Scale log probability by action dimensions
	size = eps.shape[-1] if action_dims is None else action_dims
	scaled_log_prob = log_prob * size

	# Reparameterization trick
	action = mean + eps * log_std.exp()
	mean, action, log_prob = math.squash(mean, action, log_prob)

	entropy_scale = scaled_log_prob / (log_prob + 1e-8)
	info = TensorDict({
		"mean": mean,
		"log_std": log_std,
		"action_prob": 1.,
		"entropy": -log_prob,
		"scaled_entropy": -log_prob * entropy_scale,
	})
	return action, info
```

**Comment by nicklashansen** [comment]
Thanks for your interest! It is true that our implementation differs a bit from SAC in this regard. I think reading through the discussion in issue https://github.com/nicklashansen/tdmpc2/issues/44 should provide some clarity!

**Comment by niamorg** [comment]
Thank you for your quick response, I added a comment to the issue https://github.com/nicklashansen/tdmpc2/issues/44
